
# CA 2

# Q1

# A


$$ Y(s) =  
\frac{1}{s(Ts+1)} U(s) = \frac{10}{s^2+10s}U(s) $$ 
$$ \ddot{y} = -10\dot{y} + 10u$$
$$ \because x(t) = (\dot{y(t)}, y(t))^T$$
$$ \therefore The\ continuous-time\ state-space\ model\ is:$$
$$
\dot{\begin{bmatrix}
x_1 \\
x_2 
\end{bmatrix}} = {\begin{bmatrix}
-10 & 0\\
1  & 0
\end{bmatrix}} {\begin{bmatrix}
x_1 \\
x_2 
\end{bmatrix}}+ {\begin{bmatrix}
10 \\
0 
\end{bmatrix}}u$$
$$ y = [0, 1] 
{\begin{bmatrix}
x_1 \\
x_2 
\end{bmatrix}}
$$
$$ A = {\begin{bmatrix}
-10 & 0\\
1  & 0
\end{bmatrix}},  B = {\begin{bmatrix}
10 \\
0
\end{bmatrix}}, C = [0, 1]$$
$After\ sampling,\ the\ state\ space\ changes\ into\ discrete-time\ form:$
$$x(k+1) = \Phi x(k) + \Gamma u(k) $$
$$\Phi = e^{Ah},  \Gamma =\int_{0}^{h}e^{Av}dv  $$

# B

$$\lambda_0 = -10, \lambda_1 = 0 $$
$$h(\lambda) = \beta_0 + \beta_1 \lambda $$
$$ \left\{
\begin{aligned}
e^{\lambda_0h} & = & \beta_0 + \beta_1 \lambda_0 \\
e^{\lambda_1h} & = & \beta_0 + \beta_1 \lambda_1 
\end{aligned} \right.
=\left\{
\begin{aligned}
e^{-10h} & = & \beta_0 - 10\beta_1  \\
e^{0h} & = & \beta_0 
\end{aligned}
\right. $$
$$\left\{
    \begin{matrix} 
    \beta_0 & = & 1  \\ 
    \beta_1 & = & \frac {1-e^{-10h}}{10}
    \end{matrix}
    \right.
$$
$$\Phi = e^{Ah} = \beta_0 I+ \beta_1 A = {\begin{bmatrix}
e^{-10h} & 0\\
\frac{1-e^{-10h}}{10}  & 1
\end{bmatrix}} $$
$$ \Gamma =\int_{0}^{h}e^{Av}dvB = \int_0^h  {\begin{bmatrix}
e^{-10v} & 0\\
\frac{1-e^{-10v}}{10}  & 1
\end{bmatrix}} dv 
{\begin{bmatrix}
10\\
0
\end{bmatrix}} = {\begin{bmatrix}
1-\frac{1}{10}e^{-10h}\\
h+\frac{1}{10}(-1+e^{-10h}) 
\end{bmatrix}} $$


$Assuming\ that\  u(k) = -Lx(k), using\ the\ deadbeat\ controller:$
$Use\ Ackermann's\ formula$
$$A_m(\phi) = \phi^2 + p_1\phi + p_2I \ \ (p_1\ and\ p_2\ are\ 0) $$
$$A_m(\phi) = {\begin{bmatrix}
e^{-20h}& 0\\
\frac{1}{10}(1-e^{-20h})& 1
\end{bmatrix}}$$
$$W_c = [\Phi, \Gamma \Phi]$$
$$L = [0, 1]W_c^{-1}A_m(\Phi)$$
$By \ using \ Matlab, \ the \ result \ is$
$$L = {\begin{bmatrix}
-\frac{10h+e^{10h}-e^{20h}}{10h(e^{10h}-1)^2}\\
\frac{e^{10h}}{h(e^{10h}-1)}
\end{bmatrix}}$$

$$ u(k) = -Lx(k) =-{\begin{bmatrix}
-\frac{10h+e^{10h}-e^{20h}}{10h(e^{10h}-1)^2}\\
\frac{e^{10h}}{h(e^{10h}-1)}
\end{bmatrix}} x(k) $$





# C

$$ u(0) = -Lx(0) = - \frac{5h+3e^{10h}-3e^{20h}}{5h(e^{10h}-1)^2}$$
$$ |u(0)|<1, h>0$$
$ Can't\ be\ solve\ by\ hand\ and\ Matlab$


# Q2


$$z(k) = \begin{bmatrix}
x(k)\\
v(k)
\end{bmatrix}$$

$$z(k+1) = \begin{bmatrix}
x(k+1)\\
v(k+1)
\end{bmatrix} = 
\begin{bmatrix}
0.5 & 1   & 1\\
0.2 & 0.8 & 0\\
0   & 0   & 1
\end{bmatrix}
\begin{bmatrix}
x(k)\\
v(k)
\end{bmatrix} =
\begin{bmatrix}
\Phi   & \Phi_{xw}\\
0 & \Phi_w\\
\end{bmatrix} 
\begin{bmatrix}
x(k)\\
v(k)
\end{bmatrix} =
\begin{bmatrix}
1\\
1\\
0
\end{bmatrix}u(k)
$$
$$ y(k) = \begin{bmatrix}
1 & 0 & 0
\end{bmatrix} 
\begin{bmatrix}
x_1(k)\\
x_2(k)\\
v(k)
\end{bmatrix}
$$
$The \ state\ and\ v\ can\ be\ measured,\ assuming $
$$ u(k) = -L_c Z(k) = -Lx(k) - L_wv(k)$$
$$x(k+1) = \Phi x(k)- \Gamma(Lx(k)+L_wv(k))+ \Phi_v v(k) = (\Phi - \Gamma L)x(k) + (\Phi_w-\Gamma L_w)v(k)$$
$To\ cancel\ the\ disurbance\ we\ need\ (\Phi_w-\Gamma L_w)\ to\ be\ zero$
$$\Phi_w-\Gamma L_w= 
\begin{bmatrix}
1\\
0
\end{bmatrix} - 
\begin{bmatrix}
1\\
1
\end{bmatrix} L_w = 
\begin{bmatrix}
1-L_w\\
L_w
\end{bmatrix}\ Can\ not\ to\ be\ zero
$$

$So\ need\ to\ design\ L,\ let\ all\ the (\Phi - \Gamma L)\ poles\ to\ zero$

$$zI-(\Phi - \Gamma L) = 
\begin{bmatrix}
z-0.5+l_1 & l_1-1\\
l_2-0.2   & z-0.8+l_2
\end{bmatrix}
$$
$$ |zI-(\Phi - \Gamma L)| = z^2$$
$$l_1 = 0.45, l_2 = 0.85$$
$Analyze the TF from v(k) to y(k):$
$$Y(z) = cX(z)\ ==>\ H_w(z) = c(zI-\Phi + \Gamma L)^{-1}(\Phi_{xw}-\Gamma L_w)v(z)$$
$Let\ H_w(1)\ = 0$
$$ l_w = 2.1$$
$$\therefore  u(k) = -
\begin{bmatrix}
0.45 & 0.85

\end{bmatrix}x(k)-2.1v(k)$$

# Q2 B  use pure state feedback?

$$\hat{z}(k+1) = 
\begin{bmatrix}
\Phi   & \Phi_{xw}\\
0 & \Phi_w\\
\end{bmatrix} \hat{z}(k)+
\begin{bmatrix}
\Gamma\\
0
\end{bmatrix}u(k)+K
\begin{bmatrix}
y(k)-\hat{y}(k)
\end{bmatrix}
$$

$$ \hat{y}(k) = 
\begin{bmatrix}
c & 0
\end{bmatrix} \hat{z}(k)
\begin{bmatrix}
x_1(k)\\
x_2(k)\\
v(k)
\end{bmatrix}
$$
$The \ state\ and\ v\ can\ be\ measured,\ assuming $


# Q2 C

$The \ state\ and\ v\ can\ be\ measured,\ assuming $
$$ u(k) = -L_c Z(k) = -Lx(k) - L_wv(k)$$
$$\hat{z}(k+1) = 
\begin{bmatrix}
\Phi   & \Phi_{xw}\\
0 & \Phi_w\\
\end{bmatrix} \hat{z}(k)+
\begin{bmatrix}
\Gamma\\
0
\end{bmatrix}u(k)+K
\begin{bmatrix}
y(k)-\hat{y}(k)
\end{bmatrix}
$$

$$ \hat{y}(k) = 
\begin{bmatrix}
c & 0
\end{bmatrix} \hat{z}(k)
\begin{bmatrix}
x_1(k)\\
x_2(k)\\
v(k)
\end{bmatrix}
$$

$$ e(k) =z(k)- \hat{z}(k)$$
$$ e(k+1) =(\Phi - Kc)e(k)$$
Place the poles to zero. The the K can be chosen.
$$zI - (\Phi - Kc) =
\begin{bmatrix}
k1 + z - 1/2&      -1&    -1\\
    k2 - 1/5& z - 4/5&     0\\
          k3&       0& z - 1
\end{bmatrix}
$$
$$det(zI - (\Phi - Kc)) = z^3$$
$$K = [2.3\ -2.36\ \ 5]^T$$
The controller is:
$$\therefore  u(k) = -
\begin{bmatrix}
0.45 & 0.85

\end{bmatrix}x(k)-2.1v(k)$$

# Q3

$$Magitude = e^{-\frac{\pi\zeta}{\sqrt{1-\zeta^2}}}<=5\%$$

In [35]:
import numpy as np
from sympy import *
zeta = symbols('zeta')
zeta_res = solve(exp(-pi*zeta/sqrt(1-zeta**2))-0.05, zeta)[0]
print('zeta:',zeta_res)
w0 =  4.6/6.9
print('w0:', w0)


zeta: 0.690106730559822
w0: 0.6666666666666666



$$\zeta = 0.69$$
$$t_s = \frac{4.6}{\zeta w_n} <= 10s $$
$$w_n = 0.67$$
The reference model is:
$$H_m(s) = \frac{w_n^2}{s^2 + 2\zeta w_n s+w_n^2} = \frac{w_n^2}{s^2 + 0.92s+0.43}$$


In [36]:

sampling = 0.1
w = w0*sqrt(1-zeta_res**2)
apha = exp(-zeta_res*w0 * sampling)
beta = cos(w*sampling)
gamma = sin(w*sampling)

b1 = 1-apha*(beta + zeta_res*w0*gamma/w)
b2 = apha**2 + apha*(zeta_res*w0*gamma/w - beta)
a1 = -2*apha*beta
a2 = apha**2
print(b1, b2, a1, a2)

0.00215480968809378 0.00208971642955746 -1.90784764358737 0.912092169705017


The discrete-time trasfer function is:
$$ H_m(s) = \frac{0.00215z + 0.0021}{z^2-1.9078z+0.9121}$$

# Q3 2

$$ 800 \ddot{y}(t) + 200\dot{y}(t) = u(t) $$
$$x_1(t) =y(t),\ x_2(t) = \dot{y}(t)$$
$$\left\{
    \begin{matrix} 
    \dot{x}_1 & = & x_2(t)  \\ 
    \dot{x}_2 & = & -0.2x_2 + 0.01u
    \end{matrix}
    \right. =
    \begin{bmatrix} 
    0     & 1 \\ 
    0     & -0.2
    \end{bmatrix}
    \begin{bmatrix} 
    x_1       \\ 
    x_2  
    \end{bmatrix} +
    \begin{bmatrix} 
    0       \\ 
    0.001 
    \end{bmatrix}u
$$
$$
y(t) = \begin{bmatrix}  1&0 \end{bmatrix}x(t)
$$
$$
A = \begin{bmatrix} 
    0     & 1 \\ 
    0     & -0.2
    \end{bmatrix}

B = \begin{bmatrix} 
    0       \\ 
    0.001 
    \end{bmatrix}

C = \begin{bmatrix}  1&0 \end{bmatrix}
$$
The discrete-time mode is:
$$ x(k+1) = \Phi x(k) + \Gamma u(k) $$
$$ y(k) = Cx(k) $$


In [43]:
A = [[0, 1], [0, -0.2]]
h = 0.1
expm(A)

array([[1.        , 0.90634623],
       [0.        , 0.81873075]])

$$\Phi = e^{Ah} = 

$$